src/common/proxy_wasm/ngx_proxy_wasm_properties.c

```diff
@@ -935,16 +945,29 @@ ngx_proxy_wasm_properties_set_host(ngx_proxy_wasm_ctx_t *pwctx,
         ngx_memcpy(hpn->sn.str.data, path->data, path->len);
     }
 
-    hpn->value.len = value->len;
-    hpn->value.data = ngx_pstrdup(pwctx->pool, value);
-    if (hpn->value.data == NULL) {
-        return NGX_ERROR;
+    if (value->data == NULL) {
+        new_data = NULL;
+        hpn->negative_cache = 1;
+
+    } else {
+        new_data = ngx_pstrdup(pwctx->pool, value);
+        if (new_data == NULL) {
+            return NGX_ERROR;
+        }
     }
 
+    hpn->is_const = is_const;
+    hpn->value.len = value->len;
+    hpn->value.data = new_data;
+
     if (new_entry) {
         ngx_rbtree_insert(&pwctx->host_props_tree, &hpn->sn.node);
     }
 
+    if (retrieve) {
+        value->data = hpn->value.data;
```

Comment:
To ensure we get our string lifetimes correct without doing a double roundtrip between Lua and C:

* When Wasm code calls `get_property`, we want to trigger a Lua getter.
* That Lua getter produces a Lua string, which is a pointer to Lua VM-managed data.
* The Lua getter finishes running, and then the WasmX module needs to copy that string over to the Wasm linear memory.
* However, there's a chance that when the Lua getter finishes running, the Lua GC triggers and drops the string, because we were not holding a Lua reference to it, but rather a raw C pointer (the `data` field of `ngx_str_t`).
* For this reason, we need to copy the Lua string to C memory to be held by the WasmX module. That's why the getter calls `store_c_value`. 
* But when we store the Lua string as a C value, we want our `ngx_str_t` to contain a pointer to the C-land copy of the string, not the original Lua-managed string. Hence, we ask `set_host_property` to set the host property and also `retrieve` the updated pointer.

We could also "set" then "get", but the default behavior of the Lua-accessible "get" is to call the getter if set, so we'd need something like a "rawget" of sorts to avoid a loop, and even if we had that, it would be two FFI call roundtrips, so I figured the `retrieve` business would be the least-bad option.

Alternatively, we could simplify the API by making `ngx_proxy_wasm_properties_set` always "retrieve" the copied `value` (that change would be harmless to both places where this function is currently used, namely, `ffi` and `hfuncs`).

Reply:
So should `store_c_value` when invoked from `c_prop_setter` not have this `retrieve` argument set? Or rather, not asking for a change, only double-checking that in that case, it is unnecessary. I think that is what confused me the most, why we'd need to "get" the value when we are just setting it.

Reply:
You're right, `retrieve` doesn't need to be set in the "setter" case.